### PR TITLE
Allow extra trusted bundles when generating master certs, node config, or kubeconfig

### PR DESCRIPTION
**Pull request by liggitt**
When using named certificates, if the non-public master hostname is served using a custom certificate, the CA for that certificate must be trusted by the nodes, admin.kubeconfig, loopback kubeconfig, internal OAuth config (and possibly pods, if the custom cert is used for `kubernetes.default.svc`-type hostnames)

This PR adds the following things:
* `oadm ca create-master-certs` / `openshift start`
  * takes optional `--certificate-authority` args to include custom trusted CA bundles
  * generates a `ca-bundle.crt` containing the generated signer and any custom trusted CA bundles
  * includes the generated signer and any custom trusted CA bundles in the resulting kubeconfig files
* `oadm {create-node-config,create-api-client-config,create-kubeconfig}`
  * takes multiple `--certificate-authority` params (instead of just one)
  * includes all the specified CA bundles in the generated kubeconfigs

### CODE REVIEW

**Comment by liggitt** [comment]
@sdodson @abutcher this would let you include custom CA bundles
@deads2k still needs tests, but the rest is ready for review

File: pkg/cmd/server/admin/create_nodeconfig.go [hunk 1]
```diff
@@ -155,8 +155,14 @@ func (o CreateNodeConfigOptions) Validate(args []string) error {
 	if len(o.APIServerURL) == 0 {
 		return errors.New("--master must be provided")
 	}
-	if _, err := os.Stat(o.APIServerCAFile); len(o.APIServerCAFile) == 0 || err != nil {
-		return fmt.Errorf("--certificate-authority, %q must be a valid certificate file", cmdutil.GetDisplayFilename(o.APIServerCAFile))
+	if len(o.APIServerCAFiles) == 0 {
+		return fmt.Errorf("--certificate-authority must be a valid certificate file")
+	} else {
+		for _, caFile := range o.APIServerCAFiles {
+			if _, err := os.Stat(caFile); len(caFile) == 0 || err != nil {
```
Comment (deads2k) [inline pkg/cmd/server/admin/create_nodeconfig.go]: Do an actual parse?  I just noticed you did it in the `create_mastercerts.go` file.

File: pkg/cmd/server/admin/create_nodeconfig.go [hunk 2]
```diff
@@ -193,11 +199,16 @@ func (o CreateNodeConfigOptions) Validate(args []string) error {
 	return nil
 }
 
-func CopyFile(src, dest string, permissions os.FileMode) error {
-	// copy the cert and key over
-	if content, err := ioutil.ReadFile(src); err != nil {
-		return err
-	} else if err := ioutil.WriteFile(dest, content, permissions); err != nil {
+func CopyFile(srcFiles []string, dest string, permissions os.FileMode) error {
```
Comment (deads2k) [inline pkg/cmd/server/admin/create_nodeconfig.go]: This looks like the wrong API.  Don't you need newlines between the file contents anyway?  That would make this even stranger.

**Comment by liggitt** [comment]
[testonlyextended][extended:alternate_certs]

File: pkg/cmd/server/start/master_args.go [hunk 3]
```diff
@@ -84,6 +86,8 @@ func BindMasterArgs(args *MasterArgs, flags *pflag.FlagSet, prefix string) {
 
 	flags.StringVar(&args.EtcdDir, prefix+"etcd-dir", "openshift.local.etcd", "The etcd data directory.")
 
+	flags.StringSliceVar(&args.APIServerCAFiles, prefix+"certificate-authority", args.APIServerCAFiles, "Optional files containing signing authorities to use (in addition to the generated signer) to verify the API server's serving certificate.")
```
Comment (liggitt) [inline pkg/cmd/server/start/master_args.go]: not sure how I feel about this showing up here... if you want a config (which you need for namedCertificates), you probably want to run `start master --write-config` as a starting point, but then it feels silly to *also* have to run `create-master-certs` just to get additional certificate authorities in the kubeconfig files

Comment (deads2k) [inline pkg/cmd/server/start/master_args.go]: > not sure how I feel about this showing up here...

I think I'm ok with it.   The level of pain involved in regenerating certs is high enough to warrant having either this or a separate `write-config` command.

**Comment by deads2k** [comment]
The update for the SA automounted CA happens based on the other bits you've touched?

**Comment by deads2k** [comment]
one question, then lgtm

**Comment by liggitt** [comment]
for the automounted SA (and the loopback oauth), `--write-config` will set up a config that references the `ca-bundle.crt`. The token controller will update the CA key in the secrets if the bytes are different.

**Comment by liggitt** [comment]
[test][extended:alternate_certs]

**Comment by liggitt** [comment]
[merge]

**Pull request merged**